lib/fsm.ex

```diff
@@ -0,0 +1,93 @@
+defmodule Fsm do
+  @moduledoc """
+  action: :at_the_stop or :on_the_way
+  door: :opened or :closed
+  route: []
+  state: {action, route, door_state, open_door_request}
+  {on_the_way, _, closed, requested} -> {at_the_stop, _, opened, not_requested}
+  {on_the_way, _, closed, not_requested} -> self
+  {on_the_way, _, closed, not_requested} -> {on_the_way, _, closed, requested}
+  {at_the_stop, _, closed, requested} -> {at_the_stop, _, opened, not_requested}
+  {at_the_stop, _, opened, requested} -> self
+  {at_the_stop, _, closed, not_requested} -> {at_the_stop, _, opened, not_requested}
+  {at_the_stop, _, opened, not_requested} -> self {at_the_stop, _, closed, not_requested} -> {on_the_way, _, closed, not_requested}
+  """
+  use GenServer
+  require ForTests
+
+  def start_link([]) do
+    GenServer.start_link(__MODULE__, [])
+  end
+
+  @impl true
+  def init([]) do
+    {:ok, []}
+  end
+
+  @impl true
+  def handle_call({:set_route, initial_route}, _from, _) do
+    {:reply, :ok, {:at_the_stop, initial_route, :closed, :not_requested}}
+  end
+
+  @impl true
+  def handle_call(:get_route_and_stop, _from, {_, route, door, open_request}) do
+    {:reply, {:ok, route}, {:at_the_stop, [], door, open_request}}
+  end
+
+  @impl true
+  def handle_info(:next, state = {_, [], _, _}) do
+    ForTests.test do
+      send(:test, :done)
+    end
+    {:stop, :normal, state}
+  end
+
+  @impl true
+  def handle_info(
+    :next, {_, [head | route], door_state, open_request})
+      when head == :stop and door_state == :closed and open_request == :not_requested do
+    ForTests.test do
+      test_stuff()
+    end
+    next()
+    {:noreply, {:at_the_stop, route, :closed, :not_requested}}
+  end
+
+  @impl true
+  def handle_info(:next, {_, [head | route], _, _}) when head == :stop do
+    ForTests.test do
+      test_stuff()
+    end
+    next()
+    {:noreply, {:at_the_stop, route, :opened, :not_requested}}
+  end
+
+  @impl true
+  def handle_info(:next, {_, [head | route], _, _}) when head == :route do
+    ForTests.test do
+      test_stuff()
+    end
+    next()
+    {:noreply, {:on_the_way, route, :closed, :not_requested}}
+  end
+
+  @impl true
+  def handle_cast(:go, state = {action, _, _, _}) when action == :at_the_stop do
+    IO.puts("go")
+    next()
+    {:noreply, state}
+  end
+
+  @impl true
+  def handle_cast(:open_door_request, {a, b, c, _}) do
```

Comment:
Серьезно? `a, b, c`? Вот представьте себе: мне нужно отладить этот код, вы в отпуске, я открываю редактор…
```suggestion
  def handle_cast(:open_door_request, state) do
    state |> Tuple.delete_at(3) |> Tuple.append(:requested)
```

Или уж потрудитесь там написать `{action, route, door_state, _}`.